File: app/src/main/java/com/karthik/myweather/ui/view/BaseFragment.java
package com.karthik.myweather.ui.view;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.karthik.myweather.R;
import com.karthik.myweather.di.ViewModelFactory;
import com.karthik.myweather.ui.viewModel.BaseViewModel;

import javax.inject.Inject;

import dagger.android.support.DaggerFragment;

public abstract class BaseFragment extends DaggerFragment {

    private static final String Progress_Dialog_Tag = "Progress_Dialog_Tag";

    @Inject
    protected ViewModelFactory viewModelFactory;

    protected NavController navController;
    private AlertDialog errorDialog;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        return super.onCreateView(inflater, container, savedInstanceState);
    }

    protected void registerForErrorDialog(BaseViewModel viewModel) {
        viewModel.error.observe(this, error -> {
            showErrorDialog();
        });
    }

    protected void registerForLoadingIndicator(BaseViewModel viewModel) {
        viewModel.isLoading.observe(this, isLoading ->{
            if(isLoading){
                showProgressDialog();
            }else {
                dismissProgressDialog();
            }
        });
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        navController = Navigation.findNavController(getActivity(), R.id.fragment_navHost);
    }

    protected void showProgressDialog(){
        Fragment prev = getActivity().getSupportFragmentManager()
                .findFragmentByTag(Progress_Dialog_Tag);
        if(prev==null){
            new ProgressDialogFragment().show(
                    getActivity().getSupportFragmentManager(), Progress_Dialog_Tag);
        }
    }

    protected void dismissProgressDialog(){
        Fragment prev = getActivity().getSupportFragmentManager()
                .findFragmentByTag(Progress_Dialog_Tag);
        if (prev != null) {
            DialogFragment df = (DialogFragment) prev;
            df.dismiss();
        }
    }

    protected void showErrorDialog(){
        errorDialog = new AlertDialog.Builder(getActivity())
                .setTitle("Error")
                .setMessage("An Unknown error occured")
                .setCancelable(true)
                .setNeutralButton("OK", (dialogInterface, i) -> {
                    errorDialog.dismiss();
                })
                .show();
    }

    protected void setTitle(String title) {
        ((AppCompatActivity) getActivity()).getSupportActionBar()
                .setTitle(title);
    }

}

File: app/src/main/java/com/karthik/myweather/ui/view/BaseFragment.kt
package com.karthik.myweather.ui.view

import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.Observer
import androidx.navigation.NavController
import androidx.navigation.Navigation
import com.karthik.myweather.R
import com.karthik.myweather.di.ViewModelFactory
import com.karthik.myweather.ui.viewModel.BaseViewModel
import dagger.android.support.DaggerFragment
import javax.inject.Inject

abstract class BaseFragment : DaggerFragment() {
    @JvmField
    @Inject
    var viewModelFactory: ViewModelFactory? = null

    @JvmField
    protected var navController: NavController? = null

    protected fun registerForErrorDialog(viewModel: BaseViewModel) {
        viewModel.error.observe(this, Observer { showErrorDialog() })
    }

    protected fun registerForLoadingIndicator(viewModel: BaseViewModel) {
        viewModel.isLoading.observe(this, Observer { isLoading: Boolean ->
            if (isLoading) {
                showProgressDialog()
            } else {
                dismissProgressDialog()
            }
        })
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        navController = Navigation.findNavController(requireActivity(), R.id.navHostFragment)
    }

    private fun showProgressDialog() {
        val prev = requireActivity().supportFragmentManager
                .findFragmentByTag(Progress_Dialog_Tag)
        if (prev == null) {
            ProgressDialogFragment().show(
                    requireActivity().supportFragmentManager, Progress_Dialog_Tag)
        }
    }

    private fun dismissProgressDialog() {
        val prev = requireActivity().supportFragmentManager
                .findFragmentByTag(Progress_Dialog_Tag)
        val df = prev as DialogFragment
        df.dismiss()
    }

    private fun showErrorDialog() {
        AlertDialog.Builder(requireActivity())
                .setTitle("Error")
                .setMessage("An Unknown error occured")
                .setCancelable(true)
                .setNeutralButton("OK") { dialogInterface, _ -> dialogInterface.dismiss() }
                .show()
    }

    protected fun setTitle(title: String?) {
        (activity as AppCompatActivity?)!!.supportActionBar?.title = title
    }

    companion object {
        private const val Progress_Dialog_Tag = "Progress_Dialog_Tag"
    }
}